rust/arrow/src/compute/util.rs

```diff
@@ -44,10 +45,58 @@ where
     }
 }
 
+/// Takes/filters a list array's inner data using the offsets of the list array.
+///
+/// Where a list array has indices `[0,2,5,10]`, taking indices of `[2,0]` returns
+/// an array of the indices `[5..10, 0..2]` and offsets `[0,5,7]` (5 elements and 2
+/// elements)
+pub(super) fn take_value_indices_from_list(
+    values: &ArrayRef,
+    indices: &UInt32Array,
+) -> (UInt32Array, Vec<i32>) {
+    // TODO benchmark this function, there might be a faster unsafe alternative
+    // get list array's offsets
+    let list: &ListArray = values.as_any().downcast_ref::<ListArray>().unwrap();
+    let offsets: Vec<u32> = (0..=list.len())
+        .map(|i| list.value_offset(i) as u32)
+        .collect();
+    let mut new_offsets = Vec::with_capacity(indices.len());
+    let mut values = Vec::new();
+    let mut current_offset = 0;
+    // add first offset
+    new_offsets.push(0);
+    // compute the value indices, and set offsets accordingly
+    for i in 0..indices.len() {
+        if indices.is_valid(i) {
+            let ix = indices.value(i) as usize;
+            let start = offsets[ix];
+            let end = offsets[ix + 1];
+            current_offset += (end - start) as i32;
+            new_offsets.push(current_offset);
+            // type annotation needed to guide compiler a bit
+            let mut offsets: Vec<Option<u32>> =
+                (start..end).map(|v| Some(v)).collect::<Vec<Option<u32>>>();
+            if !offsets.is_empty() {
```

Comment:
nit: we can check this right after line 75, on whether `start = end`, so that we don't need to do the extra `map` operation.